backends-common/elasticsearch-v7/src/main/java/org/apache/james/backends/es/v7/IndexCreationFactory.java

```diff
@@ -33,65 +37,147 @@
 
-    public static class AliasSpecificationStep {
-        private final int nbShards;
-        private final int nbReplica;
-        private final int waitForActiveShards;
-        private final IndexName indexName;
-        private final ImmutableList.Builder<AliasName> aliases;
+    public static class IndexCreationCustomElement {
+        public static IndexCreationCustomElement EMPTY = from("{}");
 
-        AliasSpecificationStep(int nbShards, int nbReplica, int waitForActiveShards, IndexName indexName) {
-            this.nbShards = nbShards;
-            this.nbReplica = nbReplica;
-            this.waitForActiveShards = waitForActiveShards;
-            this.indexName = indexName;
-            this.aliases = ImmutableList.builder();
+        public static IndexCreationCustomElement from(String value) {
+            try {
+                new ObjectMapper().readTree(value);
+            } catch (JsonProcessingException e) {
+                throw new IllegalArgumentException("value must be a valid json");
+            }
+            return new IndexCreationCustomElement(value);
         }
 
-        public AliasSpecificationStep addAlias(AliasName aliasName) {
-            Preconditions.checkNotNull(aliasName);
-            this.aliases.add(aliasName);
-            return this;
-        }
+        private final String payload;
 
-        public ReactorElasticSearchClient createIndexAndAliases(ReactorElasticSearchClient client) {
-            return new IndexCreationPerformer(nbShards, nbReplica, waitForActiveShards, indexName, aliases.build()).createIndexAndAliases(client, Optional.empty());
+        IndexCreationCustomElement(String payload) {
+            this.payload = payload;
         }
 
-        public ReactorElasticSearchClient createIndexAndAliases(ReactorElasticSearchClient client, XContentBuilder mappingContent) {
-            return new IndexCreationPerformer(nbShards, nbReplica, waitForActiveShards, indexName, aliases.build()).createIndexAndAliases(client, Optional.of(mappingContent));
+        public String getPayload() {
+            return payload;
         }
     }
 
     static class IndexCreationPerformer {
+        public static class Builder {
+
+            @FunctionalInterface
+            public interface RequireNbShards {
+                RequireNbReplica nbShards(int nbShards);
+            }
+
+            @FunctionalInterface
+            public interface RequireNbReplica {
+                RequireWaitForActiveShards nbReplica(int nbReplica);
+            }
+
+            @FunctionalInterface
+            public interface RequireWaitForActiveShards {
+                RequireIndexName waitForActiveShards(int waitForActiveShards);
+            }
+
+            @FunctionalInterface
+            public interface RequireIndexName {
+                RequireAlias indexName(IndexName indexName);
+            }
+
+            @FunctionalInterface
+            public interface RequireAlias {
```

Comment:
I thought aliases would be in the FinalStage like custom analyzers and tokenizers?

Reply:
I think this is your suggestion?
Your write "I think for addAlias it is the same principle than the optional fields"

Reply:
Same principle yes... so why are they not all together in the FinalStage? :)

Reply:
I doubt with this you can add multiple times aliases